File: eksperimenter/gruppe-2_2-kokebok/frontend/src/types/Recipe.ts
export type RecipeCategory = 'FORRETT' | 'HOVEDRETT' | 'DESSERT' | 'SNACK' | 'DRIKKE'
export type FlavorTag = 'SPICY' | 'SWEET' | 'SOUR' | 'SALTY' | 'UMAMI' | 'BITTER' | 'MILD'

export interface Ingredient {
  id?: number
  sortOrder: number
  name: string
  quantity: number | null
  unit: string
}

export interface RecipeStep {
  id?: number
  stepNumber: number
  instruction: string
}

export interface Recipe {
  id: number
  title: string
  description: string
  cuisine: string
  category: RecipeCategory
  flavorTags: FlavorTag[]
  prepTimeMinutes: number | null
  cookTimeMinutes: number | null
  servings: number | null
  imageFilename: string | null
  imageUrl: string | null
  ingredients: Ingredient[]
  steps: RecipeStep[]
  createdAt: string
  updatedAt: string
}

export type RecipeInput = Omit<Recipe, 'id' | 'imageUrl' | 'createdAt' | 'updatedAt'>

export interface SearchParams {
  q?: string
  cuisine?: string
  category?: string
  ingredient?: string
  flavors?: string[]
}

export interface MenuResponse {
  dishes: Recipe[]
  totalTimeMinutes: number
}

File: eksperimenter/gruppe-2_2-kokebok/backend/src/main/java/no/scienta/kokebok/domain/Recipe.java
package no.scienta.kokebok.domain;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.*;

@Entity
@Table(name = "recipes")
public class Recipe {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotBlank
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    private String cuisine;

    @Enumerated(EnumType.STRING)
    @NotNull
    private RecipeCategory category;

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "recipe_flavor_tags", joinColumns = @JoinColumn(name = "recipe_id"))
    @Column(name = "flavor_tag")
    @Enumerated(EnumType.STRING)
    private Set<FlavorTag> flavorTags = new HashSet<>();

    private Integer prepTimeMinutes;
    private Integer cookTimeMinutes;
    private Integer servings;
    private String imageFilename;

    @CreationTimestamp
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;

    @OneToMany(mappedBy = "recipe", cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("sortOrder ASC")
    private List<Ingredient> ingredients = new ArrayList<>();

    @OneToMany(mappedBy = "recipe", cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("stepNumber ASC")
    private List<RecipeStep> steps = new ArrayList<>();

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }

    public String getTitle() { return title; }
    public void setTitle(String title) { this.title = title; }

    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }

    public String getCuisine() { return cuisine; }
    public void setCuisine(String cuisine) { this.cuisine = cuisine; }

    public RecipeCategory getCategory() { return category; }
    public void setCategory(RecipeCategory category) { this.category = category; }

    public Set<FlavorTag> getFlavorTags() { return flavorTags; }
    public void setFlavorTags(Set<FlavorTag> flavorTags) { this.flavorTags = flavorTags; }

    public Integer getPrepTimeMinutes() { return prepTimeMinutes; }
    public void setPrepTimeMinutes(Integer prepTimeMinutes) { this.prepTimeMinutes = prepTimeMinutes; }

    public Integer getCookTimeMinutes() { return cookTimeMinutes; }
    public void setCookTimeMinutes(Integer cookTimeMinutes) { this.cookTimeMinutes = cookTimeMinutes; }

    public Integer getServings() { return servings; }
    public void setServings(Integer servings) { this.servings = servings; }

    public String getImageFilename() { return imageFilename; }
    public void setImageFilename(String imageFilename) { this.imageFilename = imageFilename; }

    public LocalDateTime getCreatedAt() { return createdAt; }
    public LocalDateTime getUpdatedAt() { return updatedAt; }

    public List<Ingredient> getIngredients() { return ingredients; }
    public void setIngredients(List<Ingredient> ingredients) { this.ingredients = ingredients; }

    public List<RecipeStep> getSteps() { return steps; }
    public void setSteps(List<RecipeStep> steps) { this.steps = steps; }
}
